base_location/models/res_partner.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2016 Nicolas Bessi, Camptocamp SA
+# Copyright 2018 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    zip_id = fields.Many2one('res.city.zip', 'ZIP Location')
+
+    @api.onchange('city_id')
+    def _onchange_city_id(self):
+        if not self.zip_id:
+            super(ResPartner, self)._onchange_city_id()
+        if self.zip_id and self.city_id != self.zip_id.city_id:
+            self.zip_id = False
+            self.zip = False
+            self.city = False
+        if self.city_id:
+            return {
+                'domain': {
+                    'zip_id': [('city_id', '=', self.city_id.id)]
+                },
+            }
+        return {'domain': {'zip_id': []}}
+
+    @api.onchange('country_id')
+    def _onchange_country_id(self):
+        res = super(ResPartner, self)._onchange_country_id()
+        if self.zip_id and self.zip_id.city_id.country_id != self.country_id:
+            self.zip_id = False
+        return res
+
+    @api.onchange('zip_id')
+    def _onchange_zip_id(self):
+        if self.zip_id:
+            if self.zip_id.city_id.country_id:
+                self.country_id = self.zip_id.city_id.country_id
+            if self.zip_id.city_id.state_id:
+                self.state_id = self.zip_id.city_id.state_id
+            self.city_id = self.zip_id.city_id
+            self.zip = self.zip_id.name
+            self.city = self.zip_id.city_id.name
+
+    @api.constrains('zip_id', 'country_id', 'city_id', 'state_id')
+    def _check_zip(self):
+        for rec in self.filtered('zip_id'):
```

Comment:
do not loop twice:
```python
for rec in self:
    if not rec.zip_id:
        continue
[...]
```

Reply:
Well, this is usually running in a recordset with only one record... I understand your code has better performance but also it's less readable.

Reply:
Actually, is more readable if you see that you are explicitly discarding records that do not have a zip. With "filtered" you are kind of hiding what your doing (especially to ppl not used to odoo's shortcuts BTW).
Also, using it in loops is a bad practice because in any case you loop twice and you (or whoever takes this code as an example) are going to reuse this approach all around by default "unconsciously".